x/incentive/keeper/reward_tracker.go

```diff
@@ -0,0 +1,295 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+
+	"github.com/babylonlabs-io/babylon/x/incentive/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sdkmath "cosmossdk.io/math"
+)
+
+// AddFinalityProviderRewardsForBtcDelegations gets the current finality provider rewards
+// and adds rewards to it, without increasing the finality provider period
+// it also does not initiliaze the FP, so it must have been initialized prior
+// to adding rewards. In the sense that a FP would first receive active delegations sats
+// be properly initialized (creates current and historical reward structures in the store)
+// than will start to receive rewards for contributing.
+func (k Keeper) AddFinalityProviderRewardsForBtcDelegations(ctx context.Context, fp sdk.AccAddress, rwd sdk.Coins) error {
+	fpCurrentRwd, err := k.GetFinalityProviderCurrentRewards(ctx, fp)
+	if err != nil {
+		return err
+	}
+
+	fpCurrentRwd.AddRewards(rwd)
+	return k.setFinalityProviderCurrentRewards(ctx, fp, fpCurrentRwd)
+}
+
+// BtcDelegationActivated adds new amount of active satoshi to the delegation
+// and finality provider. Since it modifies the amount staked, it triggers
+// the creation of a new period, which initializes the FP, creates
+// historical reward tracker, withdraw the BTC delegation rewards to gauge
+// and initializes a new delegation with the just ended period.
+func (k Keeper) BtcDelegationActivated(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.addDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// BtcDelegationUnbonded it modifies the total amount of satoshi staked
+// for the delegation (fp, del) and for the finality provider by subtracting.
+// Since it modifies the active amount it triggers the increment of fp period,
+// creationg of new historical reward, withdraw of rewards to gauge
+// and initialization of a new delegation.
+// It errors out if the unbond amount is higher than the total amount staked.
+func (k Keeper) BtcDelegationUnbonded(ctx context.Context, fp, del sdk.AccAddress, sat uint64) error {
+	amtSat := sdkmath.NewIntFromUint64(sat)
+	return k.btcDelegationModifiedWithPreInitDel(ctx, fp, del, func(ctx context.Context, fp, del sdk.AccAddress) error {
+		return k.subDelegationSat(ctx, fp, del, amtSat)
+	})
+}
+
+// FpSlashed a slashed finality provider should withdraw all the rewards
+// available to it, by iterating over all the delegations for this FP
+// and sending to the gauge. After the rewards are removed, it should
+// delete every rewards tracker value in the store related to this slashed
+// finality provider.
+func (k Keeper) FpSlashed(ctx context.Context, fp sdk.AccAddress) error {
```

Comment:
Why do we need to enforce withdrawal and remove the BTC delegation reward trackers? Upon slashing this FP will lose all voting power and won't get any reward anyway. It seems secure to keep the rewards in gauge as is